### PR TITLE
Batch recommendations. Fixes #6

### LINKED ISSUE #6: Add a batch recommendation route.
Our recommendation route currently works on a 1-1 basis. But a user can recommend stuff to multiple person at a time.

### PR DESCRIPTION
**Pull request by aandis**
Also added support for creating recommendations where recommender may not be a friend of recommendee. 

TODOs - 
1. If user with the given email doesn't exist, send an email and sign him up,
2. Handle error when creating recommendations in batches.
3. Performance improvements. 

@rohitpaulk have a look.

### CODE REVIEW

File: app/models/recommendation.rb [hunk 1]
```diff
@@ -25,6 +25,52 @@ def validate(record)
 
   before_validation :set_pending_status
 
+  def self.create_by_id_and_email(recommender, params)
+    recommendee_ids = params["recommendee_ids"]
+    recommendee_emails = params["recommendee_emails"]
+    item_class = Kernel.const_get(params['item_type'])
```
Comment (rohitpaulk) [inline app/models/recommendation.rb]: Check for errors here. What if I pass in `ConstThatDoesntExist` as the `item_type`?

File: app/models/recommendation.rb [hunk 2]
```diff
@@ -25,6 +25,52 @@ def validate(record)
...
+    item = item_class.find(params['item_id'])
```
Comment (rohitpaulk) [inline app/models/recommendation.rb]: Check for errors here too. Find throws an error if the item doesn't exist.

File: app/models/recommendation.rb [hunk 3]
```diff
@@ -25,6 +25,52 @@ def validate(record)
 
   before_validation :set_pending_status
 
+  def self.create_by_id_and_email(recommender, params)
```
Comment (rohitpaulk) [inline app/models/recommendation.rb]: This should not be taking the `params` parameter. You should pass in `recommendee_ids` and `recommendee_emails` separately. Parsing params is the controller's job.

File: app/models/recommendation.rb [hunk 4]
```diff
@@ -25,6 +25,52 @@ def validate(record)
...
+    recommendee_ids = params["recommendee_ids"]
+    recommendee_emails = params["recommendee_emails"]
+    item_class = Kernel.const_get(params['item_type'])
+    item = item_class.find(params['item_id'])
+
+    new_recommendations = []
+
+    unless recommendee_ids.nil?
```
Comment (rohitpaulk) [inline app/models/recommendation.rb]: I haven't thought of a better way to do this, but this method looks verryyyy long to me. Sure you can't split it up into simpler chunks?

**Comment by aandis** [comment]
@rohitpaulk Yeah. Those checks I'll add. How should the errors be returned though? Because it's creating multiple recommendations in one api call.  Some might be successful and some might throw errors.

**Comment by rohitpaulk** [comment]
Usually what's done in this scenario is, a DB-level transaction is used. If everything succeeds, you commit the transaction. If anything fails in the middle - the transaction will be rolled back, and the error will be passed further up.

However, this depends on how important the errors are. If you're okay with only half of the recommendations being created and the rest failing, then you could do that too.

**Comment by aandis** [comment]
Although the former is simpler, the latter sounds like what we should do. Recommending to User A shouldn't be effected by a simultaneous recommendation to User B. So for each new recommendation in the call, the method would return the created recommendation or the corresponding error. Sound good?

**Comment by rohitpaulk** [comment]
> So for each new recommendation in the call, the method would return the created recommendation or the corresponding error. Sound good?

Yes, makes sense.

File: app/models/recommendation.rb [hunk 5]
```diff
@@ -25,6 +25,52 @@ def validate(record)
...
+            :item => item
+          )
+          if reco.save
+            new_recommendations.append(reco)
+          else
+            #Handle batch error
+          end
+        else
+          #TODO - New user. Send email with new recommendation.
+        end
```
Comment (aandis) [inline app/models/recommendation.rb]: Well I could refactor out the common stuff here. That oughtta make it shorter. But I don't want to go against [Rule 1](http://nathanmarz.com/blog/suffering-oriented-programming.html) for now :wink:

Comment (rohitpaulk) [inline app/models/recommendation.rb]: I'm only merging this after Rule 2 is done at least ;)

**Comment by aandis** [comment]
@rohitpaulk have a look.

**Comment by rohitpaulk** [comment]
> would it be a good idea to have one array, check if the value is email or not and proceed accordingly?

No real preference here, I'm fine with anything.

**Pull request merged**